src/control/character_controller.rs

```diff
@@ -904,3 +913,164 @@ fn subtract_hit(translation: Vector<Real>, hit: &ShapeCastHit) -> Vector<Real> {
     let surface_correction = surface_correction * (1.0 + 1.0e-5);
     translation + *hit.normal1 * surface_correction
 }
+
+#[cfg(test)]
+mod test {
+    use std::default;
+
+    use crate::{control::KinematicCharacterController, prelude::*};
+
+    #[cfg(all(feature = "dim3", feature = "f32"))]
```

Comment:
I gated it behind features because it's already a big test, I fear adding support for multiple dimensions and precision would make it even less approachable